### ISSUE TITLE
PPO: Gradient clipping is applied but not mentioned in the paper

### ISSUE THREAD

**Issue by Tejas-Chakkarwar** [Tejas-Chakkarwar/pa3-drl-code-reading#11]
## Finding

SB3 clips gradient norms with a default max of 0.5 during PPO training. The paper doesn't mention gradient clipping.

## Paper Says

The PPO paper describes optimizing the combined loss (policy + value + entropy) but doesn't say anything about clipping gradients. The whole point of PPO's clipped objective is to limit update sizes through the objective itself, not through gradient manipulation.

## Code Does

After the backward pass, gradient norms get clipped before the optimizer step:

Relevant lines: https://github.com/Tejas-Chakkarwar/pa3-drl-code-reading/blob/5e00d26428f8633141b841a51be473971e2a6ea4/stable_baselines3/ppo/ppo.py#L273-L278

Default value:

https://github.com/Tejas-Chakkarwar/pa3-drl-code-reading/blob/5e00d26428f8633141b841a51be473971e2a6ea4/stable_baselines3/ppo/ppo.py#L100

```python
max_grad_norm: float = 0.5,
```

## Analysis

Gradient clipping is a standard deep learning trick to prevent exploding gradients, not specific to PPO. The default of 0.5 is pretty aggressive. The paper's approach is to use the clipped surrogate to keep updates in check, but in practice you can still get big gradients, especially with unusual reward scales or tricky environments. For a library that needs to work across a wide range of setups, adding gradient clipping makes sense as extra insurance. But it's definitely adding a constraint that the paper doesn't describe.

## Verdict

`paper-divergence`

The paper relies on the clipped objective to constrain updates. SB3 adds gradient clipping on top of that, which is a practical engineering choice not part of the algorithm.
